go-controller/pkg/ovn/ovn.go

```diff
@@ -738,6 +793,127 @@ func (oc *Controller) WatchEgressFirewall() *factory.Handler {
+					if oc.egressfirewallDNSInfo[dnsName].ipNodes[ip] == nil {
+						oc.egressfirewallDNSInfo[dnsName].ipNodes[ip] = make(map[string]struct{})
+					}
+					oc.egressfirewallDNSInfo[dnsName].ipNodes[ip][dnsObject.Name] = struct{}{}
+					ipNet = append(ipNet, net.ParseIP(ip))
+				}
+
+				err := oc.egressfirewallDNSInfo[dnsName].as.AddIPs(ipNet)
+				if err != nil {
+					klog.Errorf("Error adding IPs to addressSet for dnsName %s: %v", dnsName, err)
+				}
+
+			}
+
+		},
+		UpdateFunc: func(older, newer interface{}) {
+			newerDNS := newer.(*dnsobject.DNSObject)
+			olderDNS := older.(*dnsobject.DNSObject)
+			klog.Infof("Updating DNSObject %s", newerDNS.Name)
+
+			for newerDNSName := range newerDNS.Spec.DNSObjectEntries {
+				var ipsToAdd []net.IP    // ip addresses that are in the new object but not the old
+				var ipsToRemove []net.IP // ip addresses that are not in the new object but in the old
+
+				if _, exists := olderDNS.Spec.DNSObjectEntries[newerDNSName]; exists {
+
+					oldIPs := make(map[string]struct{})
+					newIPs := make(map[string]struct{})
+					for _, newIPAddress := range newerDNS.Spec.DNSObjectEntries[newerDNSName].IPAddresses {
+						newIPs[newIPAddress] = struct{}{}
+					}
+					for _, oldIPAddress := range olderDNS.Spec.DNSObjectEntries[newerDNSName].IPAddresses {
+						oldIPs[oldIPAddress] = struct{}{}
+					}
+
+					for oldIPAddress := range oldIPs {
+						if _, exists := newIPs[oldIPAddress]; !exists {
+							ipsToRemove = append(ipsToRemove, net.ParseIP(oldIPAddress))
+						}
+					}
+					for newIPAddress := range newIPs {
+						if _, exists := oldIPs[newIPAddress]; !exists {
+							ipsToAdd = append(ipsToAdd, net.ParseIP(newIPAddress))
+						}
+					}
+
+				} else {
+					for _, ip := range newerDNS.Spec.DNSObjectEntries[newerDNSName].IPAddresses {
+						ipsToAdd = append(ipsToAdd, net.ParseIP(ip))
+					}
+				}
+				oc.egressfirewallDNSMutex.Lock()
+				err := oc.egressfirewallDNSInfo[newerDNSName].as.AddIPs(ipsToAdd)
+				if err != nil {
+					klog.Errorf("Error adding IPs to addressSet for dnsName %s as part of updating: %v", newerDNSName, err)
+				}
+				//could have done this above but would need to hold the lock longer
+				for _, ip := range ipsToAdd {
+
+					if oc.egressfirewallDNSInfo[newerDNSName].ipNodes[ip.String()] == nil {
+						oc.egressfirewallDNSInfo[newerDNSName].ipNodes[ip.String()] = make(map[string]struct{})
+					}
+					oc.egressfirewallDNSInfo[newerDNSName].ipNodes[ip.String()][newerDNS.Name] = struct{}{}
+				}
+				for index, ip := range ipsToRemove {
+					delete(oc.egressfirewallDNSInfo[newerDNSName].ipNodes[ip.String()], newerDNS.Name)
+					//if there is another Node that is reporting this IP address do not remove it
+					if len(oc.egressfirewallDNSInfo[newerDNSName].ipNodes[ip.String()]) > 0 {
+						ipsToRemove = append(ipsToRemove[:index], ipsToRemove[index+1:]...)
```

Comment:
This will lead to a panic. The initial range works on the len(ipsToRemove).
See https://vbauerster.github.io/2017/04/removing-items-from-a-slice-while-iterating-in-go/

Reply:
making the temp for this is no different from just appending to another slice so that is how I changed it